---
title: "Genomic Data Analytics (GDA)"
author: "Giulio Caravagna (gcaravagna@units.it)"
date: "3/5/2021"
output:
  github_document
---

```{r, include = FALSE}
knitr::opts_chunk$set(
  collapse = TRUE,
  comment = "#>",
  fig.path = "man/figures/README-",
  out.width = "100%"
)
```

_MSc program in Data Science and Scientific Computing. University of Trieste, Italy_

- 3CFU - 24h, 12 lecture, 2 hours each. 50% theoretical lecture, 50% practical session (40’ each).
- GitHub: https://github.com/caravagn/GDA

## Lecturers

- Dr Giulio Caravagna, [Cancer Data Science Laboratory.](http://www.caravagnalab.org)

**Invited guest lecturers**

- Dr Alex Graudenzi, CNR.
- Dr Daniele Ramazzotti, University of Milan-Bicocca.
- Dr Salvatore Milite, University of Trieste
- Dr Riccardo Bergamin, University of Trieste

*** 

## Program

| Lecturer   | Title       |
|:----------|:-------------|
| Caravagna |  _Variant calling from bulk sequencing_ |
| Caravagna |   _Measuring aneuploidy from bulk sequencing_ |
| Caravagna |  _Integrated quality control of somatic calls_ |
| Bergamin |   _Population genetics for cancer_ |
| Caravagna |  _Tumour subclonal deconvolution_ |
| Ramazzotti |   _Somatic mutational signatures_ |
| Bergamin, Milite |   _Basics of Single-cell RNA analysis_ |
| Graudenzi |   _Longitudinal evolution from single cell_ |
| Milite |   _Count-based models for single-cell data_ |
| Caravagna |  _Evolutionary based stratifications_ |
| Caravagna |  _Population-level models_ |

***

- Course presentation: 

    - Cancer Evolution, 
    - Modern Genomics, 
    - Single-cell. 
    - Research at the CDSLab (www.caravagnalab.org).
    
*** 

### Part 1 - Somatic calling from bulk sequencing

*** 
*** 
 
**Lecture:** _Variant calling from bulk sequencing_

- (Theory) Mutation calling: 

    - Tumour matched-normal design, 
    - High-level design of GATK
    - Joint calling model 

- (Practice) Example VCF and PCAWG:

    - VCF manipulation
    - 27 PCAWG cases (mutation types, burden, etc.)

- Readings

    - (tool) Roth, Andrew, et al. "JointSNVMix: a probabilistic model for accurate detection of somatic mutations in normal/tumour paired next-generation sequencing data." Bioinformatics 28.7 (2012): 907-913.
    - (tool) Kim, Sangtae, et al. "Strelka2: fast and accurate calling of germline and somatic variants." Nature methods 15.8 (2018): 591-594.
    - (tool) Benjamin, David, et al. "Calling somatic snvs and indels with mutect2." BioRxiv (2019): 861054.
    - (tool) Rimmer, Andy, et al. "Integrating mapping-, assembly-and haplotype-based approaches for calling variants in clinical sequencing applications." Nature genetics 46.8 (2014): 912-918.
    - (tool) GATK (Broad Institute)
        - Training: www.csc.fi/en/web/training/-/gatk2019
        - Lectures: https://www.youtube.com/watch?v=sM9cQPWwvn4&list=PLjiXAZO27elDHGlQwfd06r7coiFtpPkvI

*** 

**Lecture:**  _Measuring aneuploidy from bulk sequencing_

- (Theory) Aneuploidy and Copy Number calling:

    - Motivation 
    - ASCAT model
    - Segmentation

- (Practice) Example runs with different tools: 

    - ASCAT
    - Sequenza (inspection of alternative solutions)
    - Circular Binary Segmentation 
    - Cohort-level distribution of CNAs per chromosome (length, percentage, copy state).

- Readings

  - (tool) Favero, Francesco, et al. "Sequenza: allele-specific copy number and mutation profiles from tumor sequencing data." Annals of Oncology 26.1 (2015): 64-70.
  - (tool) Van Loo, Peter, et al. "Allele-specific copy number analysis of tumors." PNAS 107.39 (2010): 16910-1691
  - (tool) Ross, Edith M., et al. "Allele-specific multi-sample copy number segmentation in ASCAT." Bioinformatics (2020).
  - (tool) Olshen, Adam B., et al. "Circular binary segmentation for the analysis of array‐based DNA copy number data." Biostatistics 5.4 (2004): 557-572.
  - (Review) Ben-David, Uri, and Angelika Amon. "Context is everything: aneuploidy in cancer." Nature Reviews Genetics 21.1   (2020): 44-62
  - (Review) Weaver, Beth AA, and Don W. Cleveland. "Does aneuploidy cause cancer?." Current opinion in cell biology 18.6 (2006): 658-667.
  - (In vivo measurements) Bolhaqueiro, Ana CF, et al. "Ongoing chromosomal instability and karyotype evolution in human colorectal cancer organoids." Nature Genetics 51.5 (2019): 824-834.
  - (coding) DNAcopy: A Package for Analyzing DNA Copy Data https://bioconductor.org/packages/release/bioc/vignettes/DNAcopy/inst/doc/DNAcopy.pdf
   - (coding)Total copy-number segmentation using CBS. https://cran.r-project.org/web/packages/PSCBS/vignettes/CBS.pdf 

*** 

**Lecture:**  _Integrated quality control of somatic calls_

- (Theory) Validating mutations, copy number and tumour purity:

    - Cancer Cell Fractions
    - CNAqc
    - Tumour in Normal contamination (ideas)
    
    
- (Practice) Quality-control of Whole Genome Sequencing data: 

    - Implement mapping of SNVs to CNA segments
    - Validate MSeq calls with CNAqc
    - Identify good and bad samples from PCAWG

- Readings

    - Househam, Jacob, William CH Cross, and Giulio Caravagna. "A fully automated approach for quality control of cancer mutations in the era of high-resolution whole genome sequencing." bioRxiv (2021).
    - Cmero, Marek, et al. "Inferring structural variant cancer cell fraction." Nature communications 11.1 (2020): 1-15.
    - Yuan, Ke, et al. "Ccube: a fast and robust method for estimating cancer cell fractions." bioRxiv (2018): 484402.


*** 

# Parte 2 - mathematical modelling e inference from bulk

*** 
*** 

**Lecture (R Bergamin):**  _Population genetics models of growth_

- (Theory) Branching processes and other models

    - ....
    - ....
    - ....
    
- (Practice) Tumour growth simulation: 

    - Synthetic tumour generation with TEMULATOR (https://t-heide.github.io/TEMULATOR/), 
    - Inspecting VAF distributions for subclones that are about to sweep, or too small to detect
    - Example tumours from CHESS (https://github.com/sottorivalab/CHESS.cpp)
    

- Readings

    - ...
    - ...

***

**Lecture:**  _Tumour subclonal deconvolution_

- (Theory) Subclonal deconvolution:

    - Tail modelling versus subclones
    - Read counts analysis
    - Multi-sample deconvolution
    
- (Practice) Deconvolution in practice

    - MOBSTER and BMix with single-sample data
    - MOBSTER and VIBER with multi-region data.

- Readings

    - Roth, Andrew, et al. "PyClone: statistical inference of clonal population structure in cancer." Nature methods 11.4 (2014): 396-398.
    - Gillis, Sierra, and Andrew Roth. "PyClone-VI: scalable inference of clonal population structures using whole genome data." BMC bioinformatics 21.1 (2020): 1-16.
    - Miller, Christopher A., et al. "SciClone: inferring clonal architecture and tracking the spatial and temporal patterns of tumor evolution." PLoS Comput Biol 10.8 (2014): e1003665.
    - Caravagna, Giulio, et al. "Subclonal reconstruction of tumors by using machine learning and population genetics." Nature Genetics 52.9 (2020): 898-907.
    - Caravagna, Giulio, et al. "The MOBSTER R package for tumour subclonal deconvolution from bulk DNA whole-genome sequencing data." BMC bioinformatics 21.1 (2020): 1-11.

***

Lecture 

    - (Theory) Phylogenetic inference {Neighbour joining, clones trees, mutation trees, SCITE}
    - Sample trees for MSeq
    - Full subclonal deconvolution with multi-region data

***

**Lecture (D Ramazzotti):**  _Mutational signatures and processes_

- (Theory) Deconvolution of signatures from SNVs :

    - ...
    - ...
    
- (Practice) Deconvolution in practice

    - ...
    - ...

- Readings

    - ...
    - ...

***

# Part 3 - Single-cell genomics transcriptomics

***
***

**Lecture (R Bergamin, S Milite):**  _Basics of Single-cell RNA analysis_

- (Theory) Single-cell RNA analysis

    - ...
    - ...
    
- (Practice) Seurat analysis

    - ...
    - ...

- Readings

    - ...
    - ...

***

Lecture 

**Lecture (A Graudenzi):**  _Longitudinal evolution from single cell_

- (Theory) Principles of evolution from single-cell

    - ...
    - ...
    
- (Practice) Inference in practice

    - ...
    - ...

- Readings

    - ...
    - ...

***

**Lecture (D Ramazzotti):**  _Somatic mutational signatures_

- (Theory) Deconvolution of signatures from SNVs :

    - ...
    - ...
    
- (Practice) Deconvolution in practice

    - ...
    - ...

- Readings

    - ...
    - ...


***

**Lecture (S Milite):**  _Count-based models for single-cell data_

- (Theory) Copy Number genotyping from single-cell RNA sequencing

    - ...
    - ...
    
- (Practice) Deconvolution in practice

    - ...
    - ...

- Readings

    - Milite, Salvatore, Riccardo Bergamin, and Giulio Caravagna. "Genotyping Copy Number Alterations from single-cell RNA sequencing." bioRxiv (2021).
    - Fan, Jean, et al. "Linking transcriptional and genetic tumor heterogeneity through allele analysis of single-cell RNA-seq data." Genome research 28.8 (2018): 1217-1227.
    - Harmanci, Akdes Serin, Arif O. Harmanci, and Xiaobo Zhou. "CaSpER identifies and visualizes CNV events by integrative analysis of single-cell or bulk RNA-sequencing data." Nature communications 11.1 (2020): 1-16.
    - Navin, Nicholas, et al. "Tumour evolution inferred by single-cell sequencing." Nature 472.7341 (2011): 90-94.

***
***

# Part 4 - Population-level inference

***

**Lecture:**  _Evolutionary based stratifications_

- (Theory) Detecting repeated evolution from multi-region bulk sequencing

    - Clone-trees and tree expansion
    - Expectation Maximisation for latent model discovery
    - Evolutionary distance and cluster
    
- (Practice) Inference in practice

    - Colorectal adenomas with REVOLVER
    - TRACERx Lung Adencarcinomas with REVOLVER

- Readings

    - Caravagna, Giulio, et al. "Detecting repeated cancer evolution from multi-region tumor sequencing data." Nature methods 15.9 (2018): 707-714.


**Lecture:**  _Population-level models_

- (Theory) Bayesian Networks models

    - Conjunctive Bayesian Networks 
    - Suppes' probabilistic causation

- (Practice) Inference in practice

    - Analysis of CODREAD with PICNIC
    - Analysis of other cbio data

- Readings

    - Beerenwinkel, Niko, Nicholas Eriksson, and Bernd Sturmfels. "Conjunctive bayesian networks." Bernoulli (2007): 893-909.
    - Gerstung, Moritz, et al. "Quantifying cancer progression with conjunctive Bayesian networks." Bioinformatics 25.21 (2009): 2809-2815.
    - Caravagna, Giulio, et al. "Algorithmic methods to infer the evolutionary trajectories in cancer progression." Proceedings of the National Academy of Sciences 113.28 (2016): E4025-E4034.
    - Ramazzotti, Daniele, et al. "CAPRI: efficient inference of cancer progression models from cross-sectional data." Bioinformatics 31.18 (2015): 3016-3026.
    - Loohuis, Loes Olde, et al. "Inferring tree causal models of cancer progression with probability raising." PloS one 9.10 (2014): e108358.

    
